### ISSUE TITLE
[noverlap] Determining the appropriate scaling ratio in noverlap settings due to sigma.js v2 rescaling

### ISSUE THREAD

**Issue by apitts** [graphology/graphology#210]
I understand that there are multiple co-ordinate systems in sigma.js v2: https://github.com/jacomyal/sigma.js/blob/v2/docs/coordinates.md

This makes the application of the noverlap layout algorithm challenging when using sigma.js v2. We can have a node at {x: -0.5, y: 0.5} and another one at {x: -0.3, y: 0.3} both with a size of 10 and they will not be close to overlapping when rendered by sigma.js v2. This is because of the rescaling. Now, if we feed this graph naively to noverlap it won't work as expected (anti-collision will be applied when it should not be). 

I believe the solution lies in the scaling ratio setting in noverlap so that we can scale these node sizes down appropriately. What is not clear though to me is how to obtain the appropriate scaling ratio from sigma.js v2. It's not the scale size: https://github.com/jacomyal/sigma.js/blob/v2/src/sigma.ts#L1516 but the scale size / zoom level is I think a component. The ratio computed here (https://github.com/jacomyal/sigma.js/blob/v2/src/utils/index.ts#L116) is I think relevant but doesn't seem to be in the public API.

This question also arises in the context of the adjustSizes setting in layout-forceatlas2 in graphology. Hopefully the above makes sense as it's a bit of a tricky one!

**Comment by Yomguithereal** [comment]
Hello @apitts. I have a solution to your problem but it is not 1. optimal nor 2. adapted to current beta version of sigma, but I point it to you because I know you should be able to make this work for you: https://github.com/medialab/nansi/blob/eac7ebef36a65262585fe27356adcfb5b9289d6c/app/components/basemap/ToolBox/Layout.tsx#L18-L48

I will find a better way in a next future, but I need to find some time to make the proper computations so that you can easily make this work with a single factor that you could compute from sigma.

**Comment by apitts** [comment]
Thank you for the above @Yomguithereal! I have experimented with the inputReducer and outputReducer approach. It certainly works better that way but, as you say, there are a few issues that remain. One of them is accounting for zoom. The other issue that I see is that the graph will stop rendering and display an error ("sigma/quadtree.insertNode: At least 5 nodes are outside the global quadtree zone. You might have a problem with the normalization function or the custom bounding box") if you zoom too far in while noverlap is running as a worker. I'm also seeing the algorithm not converge where I would expect it to converge after a small expansion.  

I will keep experimenting and, of course, appreciate any future improvements thatZ you suggest may take place.

**Comment by apitts** [comment]
@Yomguithereal The reference above was indeed helpful but there were two more pieces that I needed so adding them here in case helpful for others. The first was that, to account for zoom, I needed to divide size in the inputReducer by the cameraSize ratio: 
`Math.sqrt(this.renderer.getCamera().getState().ratio)`

The second was that the new noverlap algorithm differs from the old one that I wrote and the one that is implemented in Gephi. It does not include adjacent cells in the computation of forces. When I ran the current algorithm I kept getting stuck with collisions that didn't resolve with fairly normal data. When I edited the current implementation such that it was inline with the old implementation the layout worked as I expected. Some references for convenience: 
https://github.com/gephi/gephi/blob/master/modules/LayoutPlugin/src/main/java/org/gephi/layout/plugin/noverlap/NoverlapLayout.java
https://github.com/jacomyal/sigma.js/blob/v1/plugins/sigma.layout.noverlap/sigma.layout.noverlap.js

Here is my edited version, but it also includes a change to incorporate fixed and hidden nodes: 
https://github.com/apitts/graphology/blob/master/src/layout-noverlap/iterate.js

**Comment by Yomguithereal** [comment]
@apitts regarding your first point I remember there is a `Sigma.scaleSize` function and I am wondering whether it does the same thing. I know it's not documented but nothing is yet (I will at some point :) ).

> The second was that the new noverlap algorithm differs from the old one that I wrote and the one that is implemented in Gephi.

My intention was to replicate Gephi's algorithm so if my implementation fails to do so I would treat this as a bug. I only changed some optimization stuff here and there. I will reopen this and investigate.

**Comment by apitts** [comment]
@Yomguithereal Yes, the scaleSize logic in sigmajs is indeed where I saw that ratio being computed and why I took that approach, specifically this line: https://github.com/jacomyal/sigma.js/blob/b816a1938b568d4a407b31814ffbbb0b63a3ec1b/src/sigma.ts#L1296

That makes sense - thanks for reopening!

**Comment by Yomguithereal** [comment]
@apitts I want to try a different approach. Currently the running time of the algorithm is dominated by the fact that you need to make sure you don't apply the same collision twice, and it does negate the benefit of the grid quite a lot. I want to try to sort the points by x or y (minus the size & margin, and sorting on the dimension having the most variance), and then, iterate over the nodes in order, and checking the window of next nodes that can possibly collide with it. It means we might stumble upon nodes having a close x but irrelevant y but those are easily filtered and I expect that in most cases the size of the window should be largely less than the total number of nodes. What do you think?

**Comment by apitts** [comment]
@Yomguithereal Yes, that does indeed sound a lot more efficient!
